// asterisk-cfgphonelib/cfgPhone.js
class cfgPhone{

    constructor(options){
        this.fs = require("fs");
        this.hash = require('object-hash');
        this.MD5 = require("crypto-js/md5");
        this.path = require("path");
        this.sleep = require('system-sleep');
        Object.assign(this,
            {

                sipserver : '',
                ntpserver : '',
                phones : [],
                directory : {
                    root: '',
                    md5: 'md5/',
                    version: 'version/',
                    tftp: '/srv/tftp/',
                }
            }
            , options);
    }

    cfExist(Phone) {
        let Files = this.fs.readdirSync('templates/');
        for (let File of Files) {
            if(this.path.extname(File) == ".tmp") {  if(File.includes(Phone.template)) {  return File; } }
        }
        return null;
    }

    cfProcExist(Phone) {
        let Files = this.fs.readdirSync('templates/');
        for (let File of Files) {
            if(this.path.extname(File) == ".proc") {  if(File.includes(Phone.template)) {  return File; } }
        }
        return null;
    }

    cfCheckMD5(Phone) {
        if (this.FileExist('versions/' +  Phone.mac + ".ver")) {
            let _PhoneMD5 = this.fs.readFileSync('versions/' + Phone.mac + '.ver', 'utf8');
            if (_PhoneMD5 ==  Phone.md5) { return false;}
        } else { this.fs.writeFileSync('versions/' + Phone.mac + '.ver', Phone.md5);  }
        return true;
    }

    cfCreate(Phone, ConfigFileName, ProcFileName) {
        let _Config = this.fs.readFileSync('templates/' + ConfigFileName, 'utf8');
        let _Proc = require("./templates/" + ProcFileName);

        for (let [key, value] of Object.entries(Phone)) {
            _Config = this.replaceAll(_Config, '{!'+ key.substring(1) + '}', _Proc(key.substring(1),value));
        }
        let CurrentDataTime = new Date();

        _Config = this.replaceAll(_Config, '{!version}', _Proc('version',
            CurrentDataTime.getFullYear() + '.'
            +   CurrentDataTime.getMonth() + '' + (CurrentDataTime.getDate() + 1) + '.'
            +   CurrentDataTime.getHours() + '' + CurrentDataTime.getMinutes() + '' + CurrentDataTime.getSeconds()));
        _Config = this.replaceAll(_Config, '{!ntpserver}', _Proc('ntpserver',this.ntpserver));
        _Config = this.replaceAll(_Config, '{!sipserver}', _Proc('sipserver',this.sipserver));

        this.fs.writeFileSync('versions/' + _Proc("{!mac}", Phone.mac)+ ".ver", Phone.md5);
        let _ConfigOutFileName = ConfigFileName.replace('[!' + Phone.template + ']', _Proc("{!mac}", Phone.mac));
        _ConfigOutFileName  = _ConfigOutFileName .replace('.tmp', '');
        this.fs.writeFileSync(this.directory.tftp  + _ConfigOutFileName , _Config);
    }

    cfInit(Phone)
    {

        let _Phone  = new cfgPhoneNumber(Phone);
        let ConfigFileName = this.cfExist(_Phone);
        let ProcFileName = this.cfProcExist(_Phone);
        if(ConfigFileName != null && ProcFileName!= null) {
            if (this.cfCheckMD5(_Phone)) { this.cfCreate(_Phone,ConfigFileName,ProcFileName);  }
        }
    }

    escapeRegExp(string) { return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'); }
    replaceAll(str, find, replace) { return str.replace(new RegExp(this.escapeRegExp(find), 'g'), replace); }

    ParseAsteriskCFG() {
        let Start = [];
        this.phones.forEach(PhonesConfigFile => {
                let PhonesConfig = this.fs.readFileSync(PhonesConfigFile, 'utf8');
                let PhonesConfigMD5 = this.MD5(PhonesConfig).toString();
                let PhonesConfigFileMD5;
                let IsReadConfig = true;

                if(this.FileExist('md5/' + this.path.basename(PhonesConfigFile) + '.md5')) {
                    PhonesConfigFileMD5 = this.fs.readFileSync('md5/' + this.path.basename(PhonesConfigFile) + '.md5', 'utf8');
                    if( PhonesConfigFileMD5 == PhonesConfigMD5) { IsReadConfig = false; }
                } else { this.fs.writeFileSync('md5/' + this.path.basename(PhonesConfigFile) + '.md5', PhonesConfigMD5); }

                if(IsReadConfig) {
                    PhonesConfig.split(/\r?\n/).forEach(Line => {
                            let IsStart = /\[([0-9]+)\](\((.*)\))?/.exec(Line);
                            let IsNewBlock = /\[([0-9]+)\](\((.*)\))?/.exec(Line);
                            if (IsStart != null || (IsStart == null && IsNewBlock != null)) {
                                if (Object.keys(Start).length > 0) {
                                    this.cfInit(Start);
                                    Start = [];
                                }
                                Start['number'] = IsStart[1];
                                Start['template'] = IsStart[2];
                            } else {
                                let LineParse = Line.split('=');
                                if (LineParse.length > 1) {
                                    switch (LineParse[0].toLowerCase().trim()) {
                                        case ";mac":
                                            Start['mac'] = LineParse[1].trim();
                                            break;
                                        case ";template":
                                            Start['template'] = LineParse[1].trim();
                                            break;
                                        case "secret":
                                            Start['secret'] = LineParse[1].trim();
                                            break;
                                        case "fullname":
                                            Start['fullname'] = this.replaceAll(LineParse[1].trim(),'"','');
                                            break;
                                        case "callerid":
                                            Start['callerid'] = this.replaceAll(LineParse[1].trim(),'"','');
                                            break;
                                    }
                                }
                            }
                        }
                    );
                }
            }
        );
    }

    FileExist(file) {
        let flag = true;
        try{  this.fs.accessSync(file, this.fs.constants.F_OK); }catch(e){ flag = false; }
        return flag;
    }

    Run()  {
        while(true) {
            this.ParseAsteriskCFG();
            this.sleep(2000);
        }
    }
}
module.exports = cfgPhone;
